transaction_verifier.go

```diff
@@ -0,0 +1,148 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"hash"
+
+	"github.com/algorand/go-algorand-sdk/client/v2/common/models"
+	"github.com/algorand/go-algorand-sdk/stateproofs/stateprooftypes"
+	"github.com/algorand/go-algorand-sdk/types"
+)
+
+var (
+	ErrProofLengthTreeDepthMismatch = errors.New("proof length and tree depth do not match")
+	ErrRootMismatch                 = errors.New("root mismatch")
+	ErrInvalidTreeDepth             = errors.New("invalid tree depth")
+	ErrIndexDepthMismatch           = errors.New("node index is not smaller than 2^depth")
+	ErrInvalidPosition              = errors.New("invalid position for node")
+)
+
+type Position int
+
+const (
+	left Position = iota
+	right
+)
+
+type TransactionVerifier struct {
+	genesisHash types.Digest
+}
+
+func (t *TransactionVerifier) getTransactionLeaf(txId types.Digest, stib types.Digest, hashFunc hash.Hash) stateprooftypes.GenericDigest {
+	buf := make([]byte, 2*types.DigestSize)
+	copy(buf[:], txId[:])
+	copy(buf[types.DigestSize:], stib[:])
+	leaf := append([]byte(stateprooftypes.TxnMerkleLeaf), buf...)
+	return stateprooftypes.HashBytes(hashFunc, leaf)
+}
+
+func (t *TransactionVerifier) getLightBlockHeaderLeaf(roundNumber types.Round,
+	transactionCommitment stateprooftypes.GenericDigest, seed stateprooftypes.Seed, hashFunc hash.Hash) stateprooftypes.GenericDigest {
+	lightBlockheader := stateprooftypes.LightBlockHeader{
+		RoundNumber:         roundNumber,
+		GenesisHash:         t.genesisHash,
+		Sha256TxnCommitment: transactionCommitment,
+		Seed:                seed,
+	}
+
+	lightBlockheader.ToBeHashed()
+	return stateprooftypes.HashBytes(hashFunc, lightBlockheader.ToBeHashed())
+}
+
+// getVectorCommitmentPositions takes the index provided with the proof, which is a bitmap of positions,
+// and translates it to an array of positions, starting from the LSB - this is the key difference
+// between vector commitments and merkle trees. The resulting array should contain a number of positions equal to the
+// depth of the tree - the length of the path between the root and the leaves.
+// Since each bit of the index is mapped to an element in the resulting positions array,
+// the index must contain a number of bits amounting to the depth parameter, which means it must be smaller than 2 ^ depth.
+func (t *TransactionVerifier) getVectorCommitmentPositions(index uint64, depth uint64) ([]Position, error) {
+	if depth == 0 {
```

Comment:
can this happen? I think it is already beeing checked at `computeVectorCommitmentMerkleRoot`

Reply:
I'll grant you that it's confusing, but the valid case, in which both the depth and the proof are empty, has already been checked and handled by this function's caller - and that is one of this function's assumptions. Would you like me to refactor it somehow?